### PR TITLE
[SPARK-36182][SQL] Support TimestampNTZ type in Parquet data source 

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Support TimestampNTZ type in Parquet data source. In this PR, the [timestamp types](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#timestamp) of Parquet are mapped to the two timestamp types in Spark:
<table>
    <tr colspan="3">
        <th colspan="3">Parquet type</th>
        <th>Spark Catalyst Type</th>
    </tr>
    <tr>
        <td rowspan="4">INT64 Timestamp</td>
        <td rowspan="2">isAdjustedToUTC = false</td>
        <td>unit = MILLIS</td>
        <td>TimestampNTZType</td>
    </tr>
    <tr>
        <td>unit = MICROS</td>
        <td>TimestampNTZType</td>
    </tr>
    <tr>
        <td rowspan="2">isAdjustedToUTC = true</td>
        <td>unit = MILLIS</td>
        <td>TimestampType</td>
    </tr>
    <tr>
        <td>unit = MICROS</td>
        <td>TimestampType</td>
    </tr>
    <tr>
        <td>INT96 Timestamp</td>
        <td> - </td>
        <td> - </td>
        <td> TimestampType </td>
</table>

#### Parquet writer
For TIMESTAMP_NTZ columns, it follows the Parquet Logical Type Definitions and sets the field `isAdjustedToUTC` as `false` on writing TIMESTAMP_NTZ columns. The output type is always INT64 in the MICROS time unit. Parquet's timestamp logical annotation can only be used for INT64 so that we won't support writing TIMESTAMP_NTZ as INT96. Otherwise, it is hard to decide the timestamp type on the reader side.

#### Parquet reader

- INT96 columns: The reader behavior is the same with Spark 3.2 or prior.
- Schema inference for INT64 Timestamp columns: for schema inference of one file, Spark infers TIMESTAMP_NTZ or TIMESTAMP_LTZ type according to the annotation flag `isAdjustedToUTC`.
- Row converter for INT64 Timestamp columns during reading
    - Given a TIMESTAMP_NTZ Parquet column and a catalyst schema of TIMESTAMP_LTZ type, Spark **allows** the read 
       operation since TIMESTAMP_LTZ is the "wider" type.
    - Given a TIMESTAMP_LTZ Parquet column and a catalyst schema of TIMESTAMP_NTZ type, Spark **disallows** the read operation since TIMESTAMP_NTZ is the "narrower" type.

### Why are the changes needed?

 Support TimestampNTZ type in Parquet data source 

### Does this PR introduce _any_ user-facing change?

Yes, support TimestampNTZ type in Parquet data source 

### How was this patch tested?

New UTs

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @sadikovi @cloud-fan @beliefer

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 1]
```diff
@@ -179,6 +183,21 @@ boolean isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit unit) {
       ((TimestampLogicalTypeAnnotation) logicalTypeAnnotation).getUnit() == unit;
   }
 
+  void validateTimestampType(DataType sparkType) {
+    assert(logicalTypeAnnotation instanceof TimestampLogicalTypeAnnotation);
+    // Throw an exception if the Parquet type is TimestampLTZ and the Catalyst type is TimestampNTZ.
```
Comment (gengliangwang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:188]: Here only reading TimestampLTZ as TimestampNTZ is disallowed. Suggestions are welcome.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 2]
```diff
@@ -370,6 +370,31 @@ private[parquet] class ParquetRowConverter(
...
+            .asInstanceOf[TimestampLogicalTypeAnnotation].getUnit == TimeUnit.MICROS =>
+        new ParquetPrimitiveConverter(updater)
+
+      case TimestampNTZType
+        if parquetType.asPrimitiveType().getPrimitiveTypeName == INT64 &&
+          parquetType.getLogicalTypeAnnotation.isInstanceOf[TimestampLogicalTypeAnnotation] &&
+          !parquetType.getLogicalTypeAnnotation
+            .asInstanceOf[TimestampLogicalTypeAnnotation].isAdjustedToUTC &&
+          parquetType.getLogicalTypeAnnotation
+            .asInstanceOf[TimestampLogicalTypeAnnotation].getUnit == TimeUnit.MILLIS =>
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:390]: The condition seems the same as above，could you extract them？

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala [hunk 3]
```diff
@@ -223,6 +223,12 @@ class ParquetWriteSupport extends WriteSupport[InternalRow] with Logging {
               recordConsumer.addLong(millis)
         }
 
+      case TimestampNTZType =>
+        // For TimestampNTZType column, Spark always output as INT64 with Timestamp annotation in
+        // MICROS time unit.
+        (row: SpecializedGetters, ordinal: Int) =>
+          recordConsumer.addLong(row.getLong(ordinal))
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala:230]: (row: SpecializedGetters, ordinal: Int) => recordConsumer.addLong(row.getLong(ordinal))

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 4]
```diff
@@ -158,6 +158,48 @@ abstract class ParquetQuerySuite extends QueryTest with ParquetTest with SharedS
...
+      val expected = Seq(
+        ("2016-01-01 10:11:12.123456"),
+        (null),
+        ("1965-01-01 10:11:12.123456"))
+        .toDS().select($"value".cast("timestamp_ntz"))
+      checkAnswer(sql("select * from ts"), expected)
+    }
+  }
+
+  test("SPARK-36182: can't read TimestampLTZ as TimestampNTZ") {
```
Comment (sadikovi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala:178]: Can we add a test for reading TimestampNTZ as TimestampLTZ?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 5]
```diff
@@ -179,6 +183,21 @@ boolean isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit unit) {
...
+    // This is to avoid mistakes in reading the timestamp values.
+    if (((TimestampLogicalTypeAnnotation) logicalTypeAnnotation).isAdjustedToUTC() &&
+      sparkType == DataTypes.TimestampNTZType) {
+      converterErrorForTimestampNTZ("int64 time(" + logicalTypeAnnotation + ")");
+    }
+  }
+
+  void converterErrorForTimestampNTZ(String parquetType) {
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java]: Thanks, updated

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala:178]: Thanks, updated

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 6]
```diff
@@ -179,6 +183,21 @@ boolean isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit unit) {
...
+    // This is to avoid mistakes in reading the timestamp values.
+    if (((TimestampLogicalTypeAnnotation) logicalTypeAnnotation).isAdjustedToUTC() &&
+      sparkType == DataTypes.TimestampNTZType) {
+      convertErrorForTimestampNTZ("int64 time(" + logicalTypeAnnotation + ")");
+    }
+  }
+
+  void convertErrorForTimestampNTZ(String parquetType) {
+    throw new RuntimeException("Unable to create Parquet converter for data type " +
+      DataTypes.TimestampNTZType.json() + " whose Parquet type is " + parquetType);
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:198]: can we use the SQL format? `TIMESTAMP WITHOUT TIMEZONE`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 7]
```diff
@@ -158,6 +158,71 @@ abstract class ParquetQuerySuite extends QueryTest with ParquetTest with SharedS
     }
   }
 
+  test("SPARK-36182: writing and reading TimestampNTZType column") {
+    withTable("ts") {
+      sql("create table ts (c1 timestamp_ntz) using parquet")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala:163]: can we test that we can not insert ltz into a ntz column?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:390]: We can extract unit as parameter.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala (hunk 7 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala:163]: We should forbid it in the store assignment rule, which is independent of data sources.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java (hunk 6 above)
Comment (gengliangwang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:198]: Both are "timestamp_ntz"

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:198]: I was not referring to `.sql`, I mean "TIMESTAMP WITHOUT TIMEZONE".

Comment (gengliangwang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:198]: I prefer keeping "timestamp_ntz", which is the keyword we used in DDL/literals..

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:198]: nvm, `timestamp_ntz` is kind of an alias of `TIMESTAMP WITHOUT TIMEZONE` in Spark today, and a shorter name is better here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala (hunk 2 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:382]: What do you mean? Isn't it what the code is already doing?

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**